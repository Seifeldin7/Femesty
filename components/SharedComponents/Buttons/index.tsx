export * from './Button'
export * from './ImageButton'
export * from './SocialButton'