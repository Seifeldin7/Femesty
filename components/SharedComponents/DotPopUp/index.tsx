import React from 'react';
import {
    View,
    Animated,
} from 'react-native';
import { componentProps } from './types';
import { styles } from "./styles";
import { TouchableOpacity } from 'react-native-gesture-handler';
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome';
import { faCircle } from '@fortawesome/free-solid-svg-icons';


export const DotPopUp = ({ arrowDirection, visibility, children, onCalloutPressed }: componentProps) => {
    const renderCallout = () => {
        if (arrowDirection == 'down') {
            return (

                <Animated.View style={[styles.container,
                {
                    opacity: visibility,
                }
                ]}>
                    <FontAwesomeIcon
                        size={20}
                        icon={faCircle} style={styles.dot} />
                    <View style={styles.calloutTriangle}>
                    </View>

                    <TouchableOpacity activeOpacity={1.0} onPress={onCalloutPressed}>
                        {children}
                    </TouchableOpacity>
                </Animated.View>


            );
        }
        else if (arrowDirection == 'up') {
            return (

                <Animated.View style={[styles.container,
                {
                    opacity: visibility,
                    marginTop: -70
                }
                ]}>
                    <TouchableOpacity activeOpacity={1.0} onPress={onCalloutPressed} >
                        {children}
                    </TouchableOpacity>

                    <View style={[styles.calloutTriangle, styles.transformTriangleDown]}>
                    </View>
                    <FontAwesomeIcon
                        size={20}
                        icon={faCircle} style={styles.dot} />
                </Animated.View>

            );
        }
        else if (arrowDirection == 'right') {
            return (
                <Animated.View style={[styles.container, styles.flexDirectionRow,
                {
                    opacity: visibility,
                    marginRight: -180,
                }
                ]}>
                    <View style={[styles.calloutTriangle, styles.transformTriangleRight]}>
                    </View>

                    <TouchableOpacity activeOpacity={1.0} onPress={onCalloutPressed} >
                        <View style={styles.calloutSquare}>
                            {children}
                        </View>

                    </TouchableOpacity>
                </Animated.View>

            );
        }
        else {
            return (
                <View >
                    <FontAwesomeIcon
                        size={20}
                        icon={faCircle} style={styles.dot} />

                    <Animated.View style={[styles.container, styles.flexDirectionRow,
                    {
                        opacity: visibility,
                        marginLeft: -180,
                    }
                    ]}>
                        <TouchableOpacity activeOpacity={1.0} onPress={onCalloutPressed} >
                            <View style={styles.calloutSquare}>
                                {children}
                            </View>

                        </TouchableOpacity>

                        <View style={[styles.calloutTriangle, styles.transformTriangleLeft]}>
                        </View>
                    </Animated.View>
                </View>

            );
        }
    }

    return (
        <View testID="Callout">
            {renderCallout()}
        </View>
    )

}

