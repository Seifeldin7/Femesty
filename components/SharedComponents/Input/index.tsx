import React, { useReducer, useEffect } from 'react';
import { View, Text, TextInput } from 'react-native';
import { AnyAction } from 'redux';
import { MainColors } from '../../../styles/Colors';
import { styles } from './styles';

const INPUT_CHANGE = 'INPUT_CHANGE';
const INPUT_BLUR = 'INPUT_BLUR';
const INPUT_FOCUS = 'INPUT_FOCUS';

const inputReducer = (state: any, action: AnyAction) => {
  switch (action.type) {
    case INPUT_CHANGE:
      return {
        ...state,
        value: action.value,
        isValid: action.isValid
      };
    case INPUT_BLUR:
      return {
        ...state,
        touched: true
      };
    case INPUT_FOCUS:
      return {
        ...state,
        focused: true
      };
    default:
      return state;
  }
};

export const Input = (props: any) => {
  const [inputState, dispatch] = useReducer(inputReducer, {
    value: props.initialValue ? props.initialValue : '',
    isValid: props.initiallyValid,
    touched: false,
    focused: false
  });

  const { onInputChange, id } = props;

  useEffect(() => {
    if (inputState.touched) {
      onInputChange(id, inputState.value, inputState.isValid);
    }
  }, [inputState, onInputChange, id]);

  const textChangeHandler = (text: string) => {
    let isValid = true;
    if (props.required && text.trim().length === 0) {
      isValid = false;
    }
    if (props.minLength != null && text.length < props.minLength) {
      isValid = false;
    }
    dispatch({ type: INPUT_CHANGE, value: text, isValid: isValid });
  };

  const lostTouchHandler = () => {
    dispatch({ type: INPUT_BLUR });
  };

  const focusHandler = () => {
    dispatch({ type: INPUT_FOCUS });
  };

  return (
    <View style={styles.formControl}>
      <Text style={styles.label}>{props.label}</Text>
      <TextInput
        {...props}
        style={{ ...styles.input, borderBottomColor: (inputState.focused) ? MainColors.Light : MainColors.Dark }}
        value={inputState.value}
        onChangeText={textChangeHandler}
        onBlur={lostTouchHandler}
        onFocus={focusHandler}
        placeholder={props.placeholder}
      />
      {!inputState.isValid && inputState.touched && (
        <View style={styles.errorContainer}>
          <Text style={styles.errorText}>{props.errorText}</Text>
        </View>
      )}
    </View>
  );
};

