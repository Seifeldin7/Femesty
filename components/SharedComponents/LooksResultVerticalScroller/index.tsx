import React, { useEffect, useState } from "react";
import { View, FlatList, RefreshControl } from "react-native";
import { HomeScreenLoading } from "../../ThemeComponents/Views/LoadingViews";
import { MHorizontalScroller } from "../../HomeScreenComponents";
import { checkConnected } from "../../../utility/CheckConnection";
import { RootState } from "../../../store/configureStore";
import { useSelector } from "react-redux";
import { NoConnection } from "../../ThemeComponents/Views/NoConnectionView";

import { styles } from './styles';
import { componentProps, LOOKS_TO_FETCH_COUNT } from './types'
import Look from "../../../entities/Look";

export const LooksResultVerticalScroller = ({
  data,
  getNextLook,
  refreshLooks,
  addFavorite,
  deleteFavorite,
  navigation,
  setSelectedResultPage
}: componentProps) => {
  const [selectedPage, setSelectedpage] = useState(0);
  const [refreshing, setRefreshing] = useState(false);
  const [loading, setLoading] = useState(false);
  const [connected, setConnected] = useState(true);
  const [scrollEnd, setScrollEnd] = useState(false);
  const isEndOfList = useSelector((state: RootState) => state.looks.isEndOfList);
  const wait = (timeout: any) => {
    return new Promise((resolve) => {
      setTimeout(resolve, timeout);
    });
  };

  const CheckConnection = () => {
    checkConnected().then((res: React.SetStateAction<boolean>) => {
      setConnected(res);
      if (!res) return;
    });
    onRefresh();
  };

  useEffect(() => {
    checkConnected().then((res: React.SetStateAction<boolean>) => {
      setConnected(res);
      if (!res) return;
    });
    setLoading(true);
    if (getNextLook) {
      getNextLook(LOOKS_TO_FETCH_COUNT).then(() => {
        setLoading(false);
      });
    }
  }, []);

  useEffect(() => {
    let loadLooksCondition =
      selectedPage == data.length - LOOKS_TO_FETCH_COUNT + 1 ||
      (selectedPage == data.length - 1 && scrollEnd);
    if (!loading && loadLooksCondition && !isEndOfList) {
      setLoading(true);
      setScrollEnd(false);
      if (getNextLook) {
        getNextLook(LOOKS_TO_FETCH_COUNT).then(() => {
          setLoading(false);
        });
      }
    }
  }, [data, selectedPage, loading === true]);

  const onRefresh = React.useCallback(async () => {
    setRefreshing(true);
    setLoading(true);
    await refreshLooks();
    if (getNextLook) {
      await getNextLook(LOOKS_TO_FETCH_COUNT);
    }
    setLoading(false);
    wait(2000).then(() => setRefreshing(false));
  }, []);

  const onScrollEnd = (e: any) => {
    let contentOffset = e.nativeEvent.contentOffset;
    let viewSize = e.nativeEvent.layoutMeasurement;
    let pageNum = Math.floor(contentOffset.y / viewSize.height);
    setSelectedpage(pageNum);
    if (setSelectedResultPage) {
      setSelectedResultPage(pageNum);
    }
    setScrollEnd(true);
  }

  const renderItem = ({ item, index }: { item: Look, index: number }) => {
    return (
      <View testID="horizontalScroller">
        <MHorizontalScroller
          look={item}
          addFavorite={addFavorite}
          deleteFavorite={deleteFavorite}
          navigation={navigation}
          index={index}
          pageNumSelected={selectedPage}
        />
        {loading && index == data.length - 1 ? (
          <View testID="Vsloading" style={styles.view}>
            <HomeScreenLoading />
          </View>
        ) : null}
      </View>
    );
  }

  return (
    <View style={{ flex: 1 }}>
      {!connected ? (
        <NoConnection onCheck={CheckConnection} />
      ) : loading && !data.length ? (
        <View style={styles.view}>
          <HomeScreenLoading />
        </View>
      ) : (
        <FlatList
          testID="looksList"
          data={data}
          keyExtractor={(item, index) => index.toString()}
          pagingEnabled={true}
          showsVerticalScrollIndicator={false}
          renderItem={renderItem}
          onMomentumScrollEnd={onScrollEnd}
          refreshControl={
            <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
          }
        />
      )}
    </View>
  );
};
