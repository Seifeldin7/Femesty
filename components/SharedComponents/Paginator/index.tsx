import React from 'react';
import { View } from 'react-native'
import { FontAwesomeIcon } from '@fortawesome/react-native-fontawesome'
import { faCircle } from '@fortawesome/free-solid-svg-icons'
import { styles } from './styles';
import { componentProps } from './types';

export const Paginator = ({ curPage, maxPages }: componentProps) => {
    let pageNumbers: any = [{}];
    for (let i = 0; i < maxPages; i++) {
        pageNumbers[i] = {};
    }
    let icons = pageNumbers.map((dot: any, index: number) => {
        return (
            <View
                key={index}
                style={(curPage == index) ? styles.dotActive : styles.dotInActive}>

            </View>
        )
    })
    return (
        <View style={styles.indicator}>
            {icons}
        </View>
    )
}

