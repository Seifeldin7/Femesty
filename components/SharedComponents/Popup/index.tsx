import React from "react";
import { View } from "react-native";
import { componentProps } from "./types";
import { styles } from "./styles";
import Modal from "react-native-modal";
import { Icon } from "../../ThemeComponents/Icon";

export const Popup = ({
  isVisible,
  setIsModalVisible,
  children,
  onCloseModal,
}: componentProps) => {
  const closeModal = () => {
    setIsModalVisible ? setIsModalVisible(false) : null;
    onCloseModal ? onCloseModal() : null;
  };

  return (
    <View testID="Popup">
      {setIsModalVisible ? (
        <Modal
          isVisible={isVisible}
          onBackdropPress={closeModal}
          onSwipeComplete={closeModal}
          swipeDirection={["down"]}
          style={{margin: 0}}
        >
          <View style={styles.mainContainer}>
            <View style={styles.popupContainer}>
              <View>
                <View style={styles.bar}></View>
                <Icon
                  size="small"
                  outerStyle={styles.close}
                  onPress={() => closeModal()}
                  iconPath={require("../../../assets/icons/small_icons/close.png")}
                />
              </View>
              {children}
            </View>
          </View>
        </Modal>
      ) : (
        <Modal
          isVisible={isVisible}
          backdropOpacity={0.0}
          onBackdropPress={closeModal}
          style={{margin: 0}}
        >
          <View style={styles.mainContainer}>
            <View style={[styles.popupContainer, styles.noClose]}>
              {children}
            </View>
          </View>
        </Modal>
      )}
    </View>
  );
};
