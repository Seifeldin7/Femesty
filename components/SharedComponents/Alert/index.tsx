import React from 'react';
import { View, Text } from 'react-native';
import { styles } from "./styles";
import { componentProps } from './types'
import Modal from 'react-native-modal';
import { BasicColors, MainColors } from '../../../styles/Colors';
import { Button } from "../Buttons";
import { AntDesign } from '@expo/vector-icons';
import { ButtonStyles } from '../../../styles/Buttons';
import { Icon } from '../../ThemeComponents/Icon';
export const CustomAlert = ({
    isVisible,
    setIsAlertVisible,
    onCloseAlert,
    title,
    description,
    type,
    callableFunction }: componentProps) => {
    const closeModal = () => {
        setIsAlertVisible ? setIsAlertVisible(false) : null;
        onCloseAlert ? onCloseAlert() : null;
    }

    return (
        <Modal
            isVisible={isVisible}
            style={styles.view}
            onBackdropPress={closeModal}>
            <View style={{ justifyContent: 'center' }}>
                <Icon size="small"
                    outerStyle={styles.close}
                    onPress={() => closeModal()} 
                    iconPath={require("../../../assets/icons/small_icons/close.png")} />
                <Text style={styles.title}>
                    {title}
                </Text>
                <Text style={styles.description}>
                    {description}
                </Text>
                {(type === 'confirm') ?
                    <View>
                        <Button
                            backgroundColor={MainColors.Light}
                            title="NO"
                            style={{ ...ButtonStyles.Medium, ...styles.button }}
                            textColor={BasicColors.White}
                            onPress={() => closeModal()}
                            enable={true}></Button>
                        <Button
                            backgroundColor={BasicColors.White}
                            title="YES"
                            style={{ ...ButtonStyles.Medium, ...styles.button }}
                            textColor={BasicColors.Black}
                            onPress={() => callableFunction()}
                            enable={true}></Button>
                    </View> : <View>
                        <Button
                            backgroundColor={MainColors.Light}
                            title="OK"
                            style={{ ...ButtonStyles.Medium, ...styles.button }}
                            textColor={BasicColors.White}
                            onPress={() => closeModal()}
                            enable={true}></Button>
                    </View>}
            </View>
        </Modal>
    )
}
