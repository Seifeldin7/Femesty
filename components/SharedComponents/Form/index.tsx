import React, { useCallback, useEffect, useReducer } from "react";
import { View, KeyboardAvoidingView, ScrollView } from "react-native";
import { componentProps } from './types'
import { Input } from "../Input";
import { styles } from "./styles";
import { AnyAction } from "redux";

const FORM_INPUT_UPDATE = 'FORM_INPUT_UPDATE';

const formReducer = (state: any, action: AnyAction) => {
  if (action.type === FORM_INPUT_UPDATE) {
    const updatedValues = {
      ...state.inputValues,
      [action.input]: action.value
    };
    const updatedValidities = {
      ...state.inputValidities,
      [action.input]: action.isValid
    };
    let updatedFormIsValid = true;
    for (const key in updatedValidities) {
      updatedFormIsValid = updatedFormIsValid && updatedValidities[key];
    }
    return {
      formIsValid: updatedFormIsValid,
      inputValidities: updatedValidities,
      inputValues: updatedValues
    };
  }
  return state;
};


export const Form = ({
  formData,
  inputProps,
  onSubmit,
  submitIsPressed }: componentProps) => {
  const inputValidities = { ...formData };
  for (let key of Object.keys(inputValidities)) {
    inputValidities[key] = formData ? true : false;
  }
  const [formState, dispatchFormState] = useReducer(formReducer, {
    inputValues: {
      ...formData
    },
    inputValidities: inputValidities,
    formIsValid: formData ? true : false
  });

  useEffect(() => {
    if (submitIsPressed) {
      if (!formState.formIsValid) {
      } else {
        onSubmit(formState.inputValues);
      }
      submitIsPressed = false;
    }
  }, [formState, submitIsPressed]);

  const inputChangeHandler = useCallback(
    (inputIdentifier, inputValue, inputValidity) => {
      dispatchFormState({
        type: FORM_INPUT_UPDATE,
        value: inputValue,
        isValid: inputValidity,
        input: inputIdentifier
      });
    },
    [dispatchFormState]
  );

  return (
    <KeyboardAvoidingView
      behavior="padding"
      keyboardVerticalOffset={100}
    >
      <ScrollView>
        <View style={styles.form}>
          {Object.keys(formData).map((key: string) => {
            return (
              <Input
                key={key}
                id={key}
                label={inputProps[key] && inputProps[key].label ? inputProps[key].label : ''}
                keyboardType={inputProps[key] && inputProps[key].keyboardType ? inputProps[key].keyboardType : 'default'}
                placeholder={inputProps[key] && inputProps[key].placeholder ? inputProps[key].placeholder : ''}
                autoCapitalize={inputProps[key] && inputProps[key].autoCapitalize ? inputProps[key].autoCapitalize : 'sentences'}
                returnKeyType={inputProps[key] && inputProps[key].returnKeyType ? inputProps[key].returnKeyType : 'next'}
                errorText={inputProps[key] && inputProps[key].errorText ? inputProps[key].errorText : ''}
                onInputChange={inputChangeHandler}
                initialValue={formData ? formData[key] : ''}
                initiallyValid={!!formData}
              />
            );
          })}
        </View>
      </ScrollView>
    </KeyboardAvoidingView>
  );
}
