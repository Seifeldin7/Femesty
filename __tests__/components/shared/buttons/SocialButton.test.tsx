import React from 'react';
import {render, fireEvent} from 'react-native-testing-library';
import { SocialButton } from "../../../../components/SharedComponents/buttons";

test('should render button in all screens', () => {
    const tree = render(
        <SocialButton
            style= {{}}
            icon={'facebook'}
            color={'blue'}
            title={'Signin with Facebook'}
            key={1}
            onPress={() => {0}}
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});


describe('Button', () => {
  it('renders the passed-in name', () => {
    const {queryByText} = render(
        <SocialButton
            style= {{}}
            icon={'facebook'}
            color={'blue'}
            title={'Signin with Facebook'}
            key={1}
            onPress={() => {0}}
        />
    );
    expect(queryByText("Signin with Facebook")).not.toBeNull();
  });
});


describe('Button', () => {
    it('renders text-if no image', () => {
        const {getByTestId} = render(
            <SocialButton
                style= {{}}
                icon={'facebook'}
                color={'blue'}
                title={'Signin with Facebook'}
                key={1}
                onPress={() => {0}}
            />
        );
        expect(getByTestId("SocailButton").props.children).toBeDefined();
    });
  });
  

describe('button pressed fired', () => {
    it('clears the message field', () => {
        const {getByText} = render(
            <SocialButton
                style= {{}}
                icon={'facebook'}
                color={'blue'}
                title={'Signin with Facebook'}
                key={1}
                onPress={() => {0}}
            />
        );
        fireEvent.press(getByText('Signin with Facebook'));
    });
});


describe('check props values assigned correctly', () => {
    it('clears the message field', () => {
        const {getByTestId} = render(
            <SocialButton
                style= {{}}
                icon={'facebook'}
                color={'blue'}
                title={'Signin with Facebook'}
                key={1}
                onPress={() => {0}}
            />
        );
        expect(getByTestId("SocailButton")).toBeDefined()
    });
});
