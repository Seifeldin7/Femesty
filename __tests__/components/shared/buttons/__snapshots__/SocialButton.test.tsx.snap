// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render button in all screens 1`] = `
<View
  style={Object {}}
>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "backgroundColor": "blue",
          "borderRadius": 5,
        },
      ]
    }
    testID="SocailButton"
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "padding": 8,
          },
          Object {
            "backgroundColor": "blue",
            "borderRadius": 5,
          },
          Object {
            "justifyContent": "center",
            "padding": 15,
            "width": 675,
          },
        ]
      }
    >
      <Text />
      <Text
        style={
          Object {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Signin with Facebook
      </Text>
    </View>
  </View>
</View>
`;
